scipy/optimize/_numdiff.py

```diff
@@ -0,0 +1,297 @@
+                f1 = f(x1, *args)
+                f2 = f(x2, *args)
+                J.append((f2 - f1) / dx)
+            else:
+                x1[i] += h[i]
+                x2[i] += 2 * h[i]
+                f1 = f(x1, *args)
+                f2 = f(x2, *args)
+                dx = x2[i] - x0[i]
+                J.append((-3 * f0 + 4 * f1 - f2) / dx)
+        J = np.array(J).T
+
+    return J
+
+
+def approx_derivatives(f, x0, method='central', rel_step=None,
+                       f0=None, args=(), bounds=(None, None)):
+    """Compute a finite difference approximation of the derivatives of a
+    vector-valued function.
+
+    If a function maps from R^n to R^m, its derivatives form m x n matrix
+    called Jacobian, where an element (i, j) is equal to the partial derivative
+    of f[i] with respect to x[j].
+
+    Parameters
+    ----------
+    f : callable
+        The function of which to estimate the derivatives. Takes and returns
+        array_like.
+    x0 : array_like
+        The point at which to estimate the derivatives.
+    method : {'central', 'forward'}, optional
+        Finite difference scheme to use.
+    rel_step : None or array_like, optional
+        Relative step size. The actual step size is computed as
+        ``h[i] = rel_step[i] * max(1, abs(x[i])``. If None, it is set to
+        eps**0.5 for `method`="center" and eps**(1/3) for `method`="forward",
+        where eps is the machine epsilon for double precision floating point
+        numbers.
+    f0 : None or array_like, optional
+        If not None it is assumed to be equal to f(x0), in this case the
+        f(x0) is not called.
+    args : tuple, optional
+        Additional arguments passed to `f`.
+    bounds : 2-tuple of array_like or None, optional
+        Lower and upper bounds on components of `x`. None means that no bounds
+        are imposed.
+
+    Returns
+    -------
+    J : ndarray
+        Finite difference approximation of derivatives in the form of Jacobian
+        matrix. If n > 1 and m > 1 then it's a 2d-array. If n = 1 and m = 1
+        then it's a scalar. Otherwise it's a 1d-array.
+
+    See Also
+    --------
+    check_derivatives : Function for checking if your implementations of
+                        function and its derivatives conform with each other.
+    """
+
+    if method not in ['forward', 'central']:
+        raise ValueError('`method` must be "forward" or "central"')
+
+    x0 = np.asarray(x0)
+    lb, ub = bounds
+    if (lb is not None and np.any(x0 < lb) or
+            ub is not None and np.any(x0 > ub)):
+        raise ValueError("`x0` violates bound constraints.")
+
+    if rel_step is None:
+        if method == 'forward':
+            rel_step = np.sqrt(np.finfo(np.float64).eps)
+        elif method == 'central':
+            rel_step = np.finfo(np.float64).eps ** (1 / 3)
+
+    if np.any(rel_step <= 0):
+        raise ValueError("`rel_step` must have positive components")
```

Comment:
Then call it `dx`, as in `(f(x + dx) - f(x)) / dx`. It is indeed simpler, so unless there's a strong recommendation in the literature...

Aside: let's move the real-time conversation to gitter (gitter.im). We are creating a huge volume of email to all the scipy devs [mea culpa]